pkg/disruptors/controller.go

```diff
@@ -96,42 +104,74 @@ func (c *agentController) InjectDisruptorAgent(ctx context.Context) error {
 // ExecCommand executes a command in the targets of the AgentController and reports any error
 func (c *agentController) ExecCommand(ctx context.Context, cmd []string) error {
 	// visit each target with the same command
-	return c.Visit(ctx, func(corev1.Pod) ([]string, error) {
-		return cmd, nil
+	return c.Visit(ctx, func(corev1.Pod) (VisitOpts, error) {
+		return VisitOpts{Command: cmd}, nil
 	})
 }
 
 // Visit allows executing a different command on each target returned by a visiting function
-func (c *agentController) Visit(_ context.Context, visitor func(corev1.Pod) ([]string, error)) error {
-	var wg sync.WaitGroup
-	// ensure errors channel has enough space to avoid blocking gorutines
-	errors := make(chan error, len(c.targets))
+func (c *agentController) Visit(ctx context.Context, visitor func(corev1.Pod) (VisitOpts, error)) error {
+	// if there are no targets, nothing to do
+	if len(c.targets) == 0 {
+		return nil
+	}
+
+	execContext, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// ensure errCh channel has enough space to avoid blocking gorutines
+	errCh := make(chan error, len(c.targets))
 	for _, pod := range c.targets {
-		wg.Add(1)
 		// attach each container asynchronously
 		go func(pod corev1.Pod) {
 			// get the command to execute in the target
-			cmd, err := visitor(pod)
+			visitOpts, err := visitor(pod)
 			if err != nil {
-				errors <- fmt.Errorf("error building command for pod %s: %w", pod.Name, err)
+				errCh <- fmt.Errorf("unable to get command for pod %q: %w", pod.Name, err)
+				return
+			}
+
+			_, stderr, err := c.helper.Exec(execContext, pod.Name, "xk6-agent", visitOpts.Command, []byte{})
+
+			// if command failed, ensure the agent execution is terminated
+			if err != nil && visitOpts.Cleanup != nil {
+				// we ignore errors because k6 was cancelled, so there's no point in reporting
+				// use a fresh context because the exec context may have been cancelled or expired
+				//nolint:contextcheck
+				_, _, _ = c.helper.Exec(context.TODO(), pod.Name, "xk6-agent", visitOpts.Cleanup, []byte{})
 			}
 
-			_, stderr, err := c.helper.Exec(pod.Name, "xk6-agent", cmd, []byte{})
 			if err != nil {
-				errors <- fmt.Errorf("error invoking agent: %w \n%s", err, string(stderr))
+				errCh <- fmt.Errorf("failed command execution for pod %q: %w \n%s", pod.Name, err, string(stderr))
+				return
 			}
 
-			wg.Done()
+			errCh <- nil
 		}(pod)
 	}
 
-	wg.Wait()
+	var err error
+	pending := len(c.targets)
+	for {
+		select {
+		case e := <-errCh:
+			pending--
+			if e != nil {
+				// cancel ongoing commands
+				cancel()
+				// Save first received error as reason for ending execution
+				err = e
```

Comment:
Probably is a leftover of some previous version of the logic that needed to wait for all the ongoing execs to end.

Reply:
The logic is: the first error should be returned, but we should still wait for all goroutines to finish to allow them to do the cleanup.  These routines will finish because the context is canceled.

If the main routine returns on the first error, how can we ensure this? 

Also, if we don't save the first error, the subsequent errors (for instance, the cancellation of the context) will be returned instead. 

Maybe I'm missing something here.

Reply:
> we should still wait for all goroutines to finish to allow them to do the cleanup

You are right, I did not notice that the same goroutine is _also_ doing the cleanup. We definitely need to wait for them to finish.